Mathlib/NumberTheory/Fermat.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2024 Moritz Firsching. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Firsching
+-/
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Algebra.Order.Ring.Basic
+import Mathlib.Algebra.Order.Star.Basic
+import Mathlib.Data.Nat.Prime.Defs
+import Mathlib.Tactic.Ring.RingNF
+
+/-!
+# Fermat numbers
+
+The Fermat numbers are a sequence of natural numbers defined as $2^{2^n} + 1$, for all natural
+numbers $n$.
+
+## Main theorems
+  `coprime_fermat_fermat`: two distinct fermat numbers are coprime.
+-/
+open Nat
+open scoped BigOperators
+
+/-- The Fermat numbers:
+the $n$-th Fermat number $F_n$ is defined as $2^{2^n} + 1$. -/
+def fermat : ℕ → ℕ := fun n ↦ 2 ^ (2 ^ n) + 1
+
+@[simp]
+theorem fermat_zero : fermat 0 = 3 := by rfl
+
+@[simp]
+theorem fermat_one : fermat 1 = 5 := by rfl
+
+@[simp]
+theorem fermat_two : fermat 2 = 17 := by rfl
+
+theorem strictMono_fermat : StrictMono fermat := by
+  apply strictMono_nat_of_lt_succ
+  simp only [fermat, add_lt_add_iff_right, Nat.pow_succ]
+  exact fun n => (pow_lt_pow_iff_right one_lt_two).mpr (by aesop)
+
+theorem two_lt_fermat {n : ℕ} : 2 < fermat n := by
+  cases n
+  · simp
+  · exact lt_of_succ_lt <| strictMono_fermat <| zero_lt_succ _
+
+theorem odd_fermat {n : ℕ} : Odd (fermat n) := by
+  rw [fermat, ← Nat.not_even_iff_odd, Nat.even_add_one, not_not, Nat.even_pow]
+  refine ⟨even_two, Nat.ne_of_gt (pow_pos zero_lt_two _)⟩
+
+@[simp]
+theorem fermat_product {n : ℕ} : ∏ k in Finset.range n, fermat k = fermat n - 2 := by
+  induction' n with n hn
+  · trivial
+  rw [Finset.prod_range_succ, hn, fermat, fermat, mul_comm,
+    (show 2 ^ 2 ^ n + 1 - 2 = 2 ^ 2 ^ n - 1 by omega),  ← Nat.sq_sub_sq]
+  ring_nf
+  omega
+
+theorem  fermat_eq_prod_add_two {n : ℕ} :  fermat n = (∏ k in Finset.range n, fermat k) + 2 := by
+  rw [fermat_product, Nat.sub_add_cancel]
+  exact le_of_lt two_lt_fermat
+
+/-- **Goldbach's theorem** : no two distinct Fermat numbers share a common factor greater
```

Comment:
I hadn't ever heard of this name for this theorem, and Google seems to have the same opinion - every single result I'm finding is for Goldbach's conjecture instead. Is this name used in any literature?

Reply:
it is mentioned in wikipedia:
https://en.wikipedia.org/wiki/Fermat_number
and here
https://proofwiki.org/wiki/Goldbach%27s_Theorem
A search for "Goldbach's theorem" fermat numbers also give many books